amr-wind/turbulence/LES/Kosovic.cpp

```diff
@@ -0,0 +1,157 @@
+Kosovic<Transport>::Kosovic(CFDSim& sim)
+    : TurbModelBase<Transport>(sim)
+    , m_vel(sim.repo().get_field("velocity"))
+    , m_rho(sim.repo().get_field("density"))
+    , m_Nij(sim.repo().declare_field("Nij", 9, 1, 1))
+    , m_divNij(sim.repo().declare_field("divNij", 3, 1, 1))
+{
+    amrex::ParmParse pp("Kosovic");
+    pp.query("Cb", m_Cb);
+    m_Cs = std::sqrt(8 * (1 + m_Cb) / (27 * M_PI * M_PI));
+    m_C1 = std::sqrt(960) * m_Cb / (7 * (1 + m_Cb) * m_Sk);
+    m_C2 = m_C1;
+    pp.query("refMOL", m_refMOL);
+    pp.query("surfaceRANS", m_surfaceRANS);
+    if (m_surfaceRANS) {
+        m_surfaceFactor = 1;
+        pp.query("switchLoc", m_switchLoc);
+        pp.query("surfaceRANSExp", m_surfaceRANSExp);
+    } else {
+        m_surfaceFactor = 0;
+    }
+    pp.query("writeTerms", m_writeTerms);
+    if (m_writeTerms) {
+        this->m_sim.io_manager().register_io_var("Nij");
+        this->m_sim.io_manager().register_io_var("divNij");
+    }
+    pp.query("LESOff", m_LESTurnOff);
+}
+
+template <typename Transport>
+void Kosovic<Transport>::update_turbulent_viscosity(
+    const FieldState fstate, const DiffusionType /*unused*/)
+{
+    BL_PROFILE(
+        "amr-wind::" + this->identifier() + "::update_turbulent_viscosity");
+
+    auto& mu_turb = this->mu_turb();
+    const auto& repo = mu_turb.repo();
+    const auto& vel = m_vel.state(fstate);
+    const auto& den = m_rho.state(fstate);
+    const auto& geom_vec = repo.mesh().Geom();
+    const amrex::Real Cs_sqr = this->m_Cs * this->m_Cs;
+
+    // Populate strainrate into the turbulent viscosity arrays to avoid creating
+    // a temporary buffer
+    fvm::strainrate(mu_turb, vel);
+    // Non-linear component Nij is computed here and goes into Body Forcing
+    fvm::nonlinearsum(m_Nij, vel);
+    fvm::divergence(m_divNij, m_Nij);
+    const int nlevels = repo.num_active_levels();
+    for (int lev = 0; lev < nlevels; ++lev) {
+        const auto& geom = geom_vec[lev];
+        const auto& problo = repo.mesh().Geom(lev).ProbLoArray();
+        const amrex::Real dx = geom.CellSize()[0];
+        const amrex::Real dy = geom.CellSize()[1];
+        const amrex::Real dz = geom.CellSize()[2];
+        const amrex::Real ds = std::cbrt(dx * dy * dz);
+        const amrex::Real ds_sqr = ds * ds;
+        const amrex::Real smag_factor = Cs_sqr * ds_sqr;
+
+        for (amrex::MFIter mfi(mu_turb(lev)); mfi.isValid(); ++mfi) {
+            const auto& bx = mfi.tilebox();
+            const auto& mu_arr = mu_turb(lev).array(mfi);
+            const auto& rho_arr = den(lev).const_array(mfi);
+            const auto& divNijLevel = (this->m_divNij)(lev).array(mfi);
+            amrex::ParallelFor(
+                bx, [=] AMREX_GPU_DEVICE(int i, int j, int k) noexcept {
+                    const amrex::Real rho = rho_arr(i, j, k);
+                    const amrex::Real x3 = problo[2] + (k + 0.5) * dz;
+                    const amrex::Real fmu = std::exp(-x3 / m_switchLoc);
```

Comment:
One key to compiling on GPUs is following the rules of for the gamma function in a ParallelFor. the [=] indicates that it will automatically capture scalars by value (e.g., dx, dy, ds_sqr), but scalars that belong to a class (e.g. m_switchLoc, m_refMOL, anything here with m_) cannot be included inside the gamma function. There is a super easy fix. Outside the ParallelFor (preferably before the lev loop), you can create local variables equal to those from a class (e.g. switchLoc = m_switchLoc). These local variables, because they don't belong to the class, can be captured and copied to device and used within the ParallelFor loop.